### ISSUE TITLE
Contradictory documentation about caching design

### ISSUE THREAD

**Issue by dustin-rcg** [marmelab/react-admin#8702]
**What you were expecting:**

I expect to the documentation to allow me to understand how the caching actually works, and I expect the documentation to be accurate and without contradiction.

**What happened instead:**

Based on [this documentation](https://marmelab.com/react-admin/Caching.html#optimistic-rendering):

This is a very heavy requirement: 

> This design choice explains why react-admin requires that all data provider methods return records of the same shape for a given resource.

It says that if I have 100 fields that can be edited for a resource, then when I get a list of that entity, say for an autocomplete on a related entity, that the list will fetch 98 fields for every entity that aren't needed. Now multiply that by the number of related entities fetched by getList for a given resource. This requirement seems like a questionable design -- indeed, my team was very surprised to read this when upgrading to v4. We are experiencing some weird caching behavior, and we need to clarify the actual design of caching by react admin.

The next sentence does not justify this requirement: 

> Otherwise, if the posts returned by getList() contain fewer fields than the posts returned by getOne(), in the previous scenario, the user will see an incomplete post at step 4.

If I ignore optimistic rendering, then based on the numbered steps, even if the entity with ID=123 is fetched and cached as part of the getList in step (2), then it will be **refetched and updated in the cache** in step (5).  In other words, the documentation says that when an entity has been cached by getList, it will be refetched and cached by getOne.  It does not say that in step (5) the entity is already in the cache and so no fetch is needed.

I don't know how this is implemented, but it seems to imply that there is a separate cache for getList and for getOne, and the optimistic rendering will try the getOne cache and will fall back to the getList cache in the case of a cache miss, but will eventually rerender when the getOne fetch is completed.

So which is it? Is the data shape physically required to be the same for all data provider methods because there is only one cache -OR- is the data shape logically required to be the same to make optimistic rendering look nice but there are separate caches for getList and getOne? These are two very different things, because if I were able to turn off optimistic rendering (or just not care about it) then I would **not** be required to use the same data shape for getList and getOne.

This seems to state that there are separate caches for the different data provider verbs:

> That means that requests for a list (getList) also populate the cache for individual records (getOne, getMany). That also means that write requests (create, udpate, updateMany, delete, deleteMany) invalidate the list cache - because after an update, for instance, the ordering of items can be changed.

And further, despite the caching

> These 3 techniques improve user experience by displaying stale data while getting fresh data from the API. But they do not reduce the ecological footprint of an app, as **the web app still makes API requests on every page**.

(Emphasis mine) It's not clear to me how the caching described in the Optimistic Rendering section relates to the Application Cache. If the Application Cache is separate (additional option), and if it is not implemented with separate caches for each data provider verb, then it would seem to defeat the design. That may be what we are experiencing.

**Comment by slax57** [comment]
Hi, and thank you for opening this enhancement request.
The fact that you are left with so many questions _after_ reading the doc clearly means that we could (and should) improve it, and try to find a clearer way of explaining how the cache works in React Admin.
Hence I'll label this issue with the documentation tag.

Now, here are some information that may be useful to understand better how it works under the hood in the meantime:
- We delegate the cache management to React-Query (v3), which is articulated around _query keys_
- In react admin, all query keys are formed in the same way: `[data provider verb, resource, params]`. Example: `['getList', 'posts', {filter: {title: 'foo'}}]`
- In the dataProvider hooks (e.g. `useGetList`), we set up React-Query to also populate the cache of other related data provider verbs. So after a successful `getList` call, the cache with query key `['getList', 'posts', {filter: {title: 'foo'}}]` will be updated, but also all caches with query keys `['getOne', 'posts', {id: 1}]`, `['getOne', 'posts', {id: 2}]`, `['getOne', 'posts', {id: 3}]` will be pre-populated
- Now, what happens when you actually call `getOne` via the `useGetOne` hook? It depends on you React-Query config.
  - By default, React-Query is configured with a `staleTime` of 0, which means all data is considered stale instantly. Hence, React-Query will first return the content from the cache (hence allowing for optimistic rendering), but will also in the background trigger an API request to fetch fresh data. Once the fresh data is available, the UI will refresh.
  - You can of course set React-Query with a `staleTime` greater than 0, in which case it will not always trigger a call to the API and will sometimes use only the cached data.
- So, in conclusion, the requirement to have records of the same shape between `getList` and `getOne` actually serves 2 purposes:
  - Avoid flickering with optimistic rendering
  - Allow to configure React-Query to return cache-only data when we need to
- But, it is fair to say that, if you leave the config to the defaults, and don't care about the flickering, then you don't need to follow this requirement at all

I hope that these explanations helped make things a bit clearer.
In any case, we should (and will) try to explain this better in the docs.

**Comment by dustin-rcg** [comment]
Thank you very much for the clear explanation.

The crux of the problem for us seems to be that during the upgrade to v4, someone set the default staleTime for the application cache. Combined with how RA does a PUT on the getOne cache in response to a getList, this appears to explain the problem we are seeing. Because the getList contains only a few fields, most of the fields of the entity are missing, and the getOne is often not getting called. Thus, the team started defining graphql fragments for all the entities and modifying all the queries to use the same fragments.

The unintended consequence seems to be that the List views are slower to load initially and in response to changing filters. The main query might take one or two seconds to execute the search in the database, but the front end is taking 5 to 9 seconds to complete the getList request. The issue seems to be the extra Data Loaders that are invoked to populate all the nested fields that aren't displayed in the list view, and they all must complete before the server can send its response, and it is also a large response. So I will discuss with the team if it made sense to introduce this setting during the upgrade. Previously we were only caching reference tables (enum tables) that are not editable in a UI. 

I feel like in our scenario, a PATCH on the getOne cache using the getList data would be a more appropriate action than a PUT.  That would avoid a flicker when subsequently navigating to the edit view to support optimistic rendering, without forcing us to query all the extra getOne data in getList. Is there any way to customize how react admin modifies the getOne cache in response to a getList, to allow us to implement custom PATCH logic? Since our list view queries are so basic compared to the edit view queries, I do not think a PATCH would hit many edge cases.

**Comment by slax57** [comment]
> Is there any way to customize how react admin modifies the getOne cache in response to a getList, to allow us to implement a custom PATCH instead of a direct PUT? Since our list view queries are so basic compared to the edit view queries, I do not think a PATCH would hit many edge cases.

I'm afraid there is not.

Indeed `useGetList` overrides the query options like this:

```ts
        {
            ...options,
            onSuccess: value => {
                const { data } = value;
                // optimistically populate the getOne cache
                data.forEach(record => {
                    queryClient.setQueryData(
                        [resource, 'getOne', { id: String(record.id), meta }],
                        oldRecord => oldRecord ?? record
                    );
                });
                // execute call-time onSuccess if provided
                if (options?.onSuccess) {
                    options.onSuccess(value);
                }
            },
        }
```

So by the time we execute the custom `onSuccess`, the `'getOne'` cache will already have been set.

Changing the `staleTime` seems like the way to go to me.

> Also, I see in the docs that it is possible to pass a staleTime value to the hooks like useGetList, which I understand would only affect the entity records fetched in that particular instance, correct? If so, then it seems like the Resource component ought to have props that let us set a default staleTime for each of the verbs on a given entity. It would just be passed through when RA automatically calls getList, for example. That would be equivalent to what we had in v3, where the caching was based on params in the data provider response. That was very useful for caching the many reference tables that aren't editable. Why isn't this supported by Resource?

Indeed I don't think there is a way to do that at the Resource level with RA v4. You will need to pass custom `queryOptions` to the components that need it such as `<List>`.

**Comment by dustin-rcg** [comment]
Thanks so much for your insightful responses. We are thinking at this point, if we set staleTime = 0, then we can add some custom caching in the data provider layer, using similar cache keys as RA for the resources that are cacheable for us.
